transport/src/main/java/io/netty/channel/ChannelHandlerAppender.java

```diff
@@ -201,5 +202,52 @@ public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
                 pipeline.remove(this);
             }
         }
+        handlerAdded0(ctx);
     }
+
+    @Override
+    public final void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        if (!selfRemoval) {
+            removeAppendHandlers(ctx.pipeline());
+        }
+        handlerRemoved0(ctx);
+    }
+
+    /**
+     * Remove all added {@link ChannelHandler} and itself (if needed) from the {@link ChannelPipeline}.
+     */
+    protected final void remove(ChannelHandlerContext ctx) {
+        ChannelPipeline pipeline = ctx.pipeline();
+        removeAppendHandlers(pipeline);
+        if (!selfRemoval) {
+            pipeline.remove(this);
+        }
+    }
+
+    private void removeAppendHandlers(ChannelPipeline pipeline) {
+        for (int i = 0; i < handlers.size(); i++) {
+            Entry h = handlers.get(i);
+            ChannelHandler handler = h.handler;
+            ChannelHandlerContext hCtx = pipeline.context(handler);
+            if (hCtx != null) {
+                try {
+                    pipeline.remove(h.handler);
+                } catch (NoSuchElementException ignore) {
+                    // A user may have removed the handler in the meantime so just ignore this if it happens.
+                }
+            }
+        }
+    }
+
+    /**
+     * Gets called after the {@link ChannelHandlerAdapter} was added to the actual context and it does handle
+     * events.
+     */
+    protected void handlerAdded0(@SuppressWarnings("unused") ChannelHandlerContext ctx) throws Exception { }
+
+    /**
+     * Gets called after the {@link ChannelHandlerAdapter} was removed from the actual context and it doesn't handle
```

Comment:
`Called after the {@link ChannelHandlerAdapter} was removed from the pipeline. At this point all handlers provided in {@link #add(ChannelHandler...)} have been removed from the pipeline.` ?